ipc/provider/src/checkpoint.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+//! Bottom up checkpoint manager
+
+use crate::config::Subnet;
+use crate::manager::{BottomUpCheckpointRelayer, EthSubnetManager};
+use anyhow::{anyhow, Result};
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use ipc_identity::{EthKeyAddress, PersistentKeyStore};
+use std::fmt::{Display, Formatter};
+use std::sync::{Arc, RwLock};
+use std::time::Duration;
+
+/// Tracks the metadata required for both top down and bottom up checkpoint submissions, such as
+/// parent/child subnet and checkpoint period.
+pub struct CheckpointMetadata {
+    parent: Subnet,
+    child: Subnet,
+    period: ChainEpoch,
+}
+
+pub struct BottomUpCheckpointManager<P, C> {
```

Comment:
I don't understand why you need two generics here `P` and `C` of type `CheckpointRelayer`. Wouldn't it be cleaner to have a single type `R` that is of type `CheckpointRelayer`. A relayer is always from a child to a parent. I don't think this abstraction is correct.

Reply:
Because the parent and child will be different network type?

Reply:
As discussed over slack, there wont be a case for supporting different network type. So a single generic should do it.